File: src/textual_webterm/docker_stats.py
"""Docker container CPU stats via Unix socket.

Reads container stats from Docker socket using only asyncio and stdlib.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import logging
import socket
from collections import deque
from pathlib import Path

log = logging.getLogger("textual-webterm")

DOCKER_SOCKET = "/var/run/docker.sock"
STATS_HISTORY_SIZE = 180  # Number of CPU readings to keep (30 min at 10s interval)
POLL_INTERVAL = 10.0  # Seconds between polls


class DockerStatsCollector:
    """Collects CPU stats from Docker containers via the Docker socket."""

    def __init__(self, socket_path: str = DOCKER_SOCKET) -> None:
        self._socket_path = socket_path
        # container_name -> deque of CPU % values (0-100)
        self._cpu_history: dict[str, deque[float]] = {}
        self._running = False
        self._task: asyncio.Task | None = None
        # Track previous CPU values for delta calculation
        self._prev_cpu: dict[str, tuple[int, int]] = {}

    @property
    def available(self) -> bool:
        """Check if Docker socket is available."""
        return Path(self._socket_path).exists()

    def get_cpu_history(self, container_name: str) -> list[float]:
        """Get CPU history for a container."""
        if container_name not in self._cpu_history:
            return []
        return list(self._cpu_history[container_name])

    async def _make_request(self, path: str) -> dict | list | None:
        """Make HTTP request to Docker socket."""
        loop = asyncio.get_event_loop()

        def _sync_request() -> bytes | None:
            try:
                sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                sock.settimeout(10.0)  # Increased timeout
                sock.connect(self._socket_path)

                # Use HTTP/1.0 to avoid chunked encoding
                request = f"GET {path} HTTP/1.0\r\nHost: localhost\r\n\r\n"
                sock.sendall(request.encode())

                # Read response
                chunks = []
                while True:
                    chunk = sock.recv(8192)
                    if not chunk:
                        break
                    chunks.append(chunk)
                sock.close()
                return b"".join(chunks)
            except (OSError, TimeoutError) as e:
                log.debug("Socket error for %s: %s", path, e)
                return None

        response = await loop.run_in_executor(None, _sync_request)
        if response is None:
            log.debug("No response from Docker socket for %s", path)
            return None

        return self._parse_docker_response(path, response)

    def _parse_docker_response(self, path: str, response: bytes) -> dict | list | None:
        """Parse HTTP response from Docker socket."""
        try:
            response_str = response.decode("utf-8", errors="replace")

            # Split headers and body
            if "\r\n\r\n" not in response_str:
                return None

            headers, body = response_str.split("\r\n\r\n", 1)

            # Check for error status
            first_line = headers.split("\r\n")[0]
            if "200" not in first_line and "OK" not in first_line:
                return None

            body = body.strip()

            # With HTTP/1.0, body should be plain JSON
            if body.startswith("{") or body.startswith("["):
                return json.loads(body)

            # Fallback: try to find JSON in body
            for line in body.split("\r\n"):
                stripped = line.strip()
                if stripped.startswith("{") or stripped.startswith("["):
                    return json.loads(stripped)

            return None
        except (json.JSONDecodeError, Exception):
            return None

    async def _discover_containers(self, service_names: list[str]) -> dict[str, str]:
        """Map service names to container IDs by querying Docker.

        Returns:
            Dict mapping service_name -> container_id
        """
        # List all containers
        containers = await self._make_request("/containers/json")
        if not isinstance(containers, list):
            return {}

        mapping: dict[str, str] = {}
        for container in containers:
            if not isinstance(container, dict):
                continue

            container_id = container.get("Id", "")[:12]  # Short ID
            names = container.get("Names", [])
            labels = container.get("Labels", {})

            # Check compose service label
            service = labels.get("com.docker.compose.service", "")
            if service in service_names:
                mapping[service] = container_id
                continue

            # Fall back to container name matching
            for name in names:
                # Docker names start with /
                clean_name = name.lstrip("/")
                # Check if service name is part of container name
                for svc in service_names:
                    if svc in clean_name or clean_name == svc:
                        mapping[svc] = container_id
                        break

        if mapping:
            log.debug("Discovered %d containers for stats", len(mapping))

        return mapping

    def _calculate_cpu_percent(
        self, container: str, cpu_stats: dict, precpu_stats: dict
    ) -> float | None:
        """Calculate CPU percentage from stats.

        Formula: (cpu_delta / system_delta) * num_cpus * 100
        """
        try:
            cpu_usage = cpu_stats.get("cpu_usage", {})
            precpu_usage = precpu_stats.get("cpu_usage", {})

            cpu_total = cpu_usage.get("total_usage", 0)
            precpu_total = precpu_usage.get("total_usage", 0)
            system_cpu = cpu_stats.get("system_cpu_usage", 0)
            presystem_cpu = precpu_stats.get("system_cpu_usage", 0)

            # Use previous values if precpu_stats is empty (first read)
            if precpu_total == 0 and container in self._prev_cpu:
                precpu_total, presystem_cpu = self._prev_cpu[container]

            # Store current values for next calculation
            self._prev_cpu[container] = (cpu_total, system_cpu)

            cpu_delta = cpu_total - precpu_total
            system_delta = system_cpu - presystem_cpu

            if system_delta <= 0 or cpu_delta < 0:
                return None

            # Get number of CPUs
            online_cpus = cpu_stats.get("online_cpus")
            if online_cpus is None:
                percpu = cpu_usage.get("percpu_usage", [])
                online_cpus = len(percpu) if percpu else 1

            cpu_percent = (cpu_delta / system_delta) * online_cpus * 100.0
            return min(cpu_percent, 100.0 * online_cpus)  # Cap at max possible

        except (KeyError, TypeError, ZeroDivisionError):
            return None

    async def _poll_container(self, service_name: str, container_id: str) -> None:
        """Poll stats for a single container."""
        path = f"/containers/{container_id}/stats?stream=false"
        stats = await self._make_request(path)

        if not isinstance(stats, dict):
            return

        cpu_stats = stats.get("cpu_stats", {})
        precpu_stats = stats.get("precpu_stats", {})

        cpu_percent = self._calculate_cpu_percent(service_name, cpu_stats, precpu_stats)
        if cpu_percent is not None:
            if service_name not in self._cpu_history:
                self._cpu_history[service_name] = deque(maxlen=STATS_HISTORY_SIZE)
            self._cpu_history[service_name].append(cpu_percent)

    async def _poll_loop(self, service_names: list[str]) -> None:
        """Background polling loop."""
        # Discover container IDs on first run and periodically refresh
        service_to_container: dict[str, str] = {}
        refresh_counter = 0
        warned_no_containers = False

        while self._running:
            # Refresh container mapping every 30 iterations (~5 minutes at 10s interval)
            if refresh_counter % 30 == 0:
                service_to_container = await self._discover_containers(service_names)
                if not service_to_container and not warned_no_containers:
                    log.warning(
                        "No Docker containers found for CPU stats. "
                        "Ensure Docker socket is mounted (-v /var/run/docker.sock:/var/run/docker.sock)"
                    )
                    warned_no_containers = True

            refresh_counter += 1

            for service_name in service_names:
                if not self._running:
                    break
                container_id = service_to_container.get(service_name)
                if not container_id:
                    continue
                try:
                    await self._poll_container(service_name, container_id)
                except Exception:
                    log.debug("Error polling stats for %s", service_name)

            await asyncio.sleep(POLL_INTERVAL)

    def start(self, service_names: list[str]) -> None:
        """Start collecting stats for given service names."""
        if not self.available:
            log.debug("Docker socket not available at %s", self._socket_path)
            return

        if self._running:
            return

        self._running = True
        self._task = asyncio.create_task(self._poll_loop(service_names))
        log.info("Started Docker stats collection for %d services", len(service_names))

    async def stop(self) -> None:
        """Stop collecting stats."""
        self._running = False
        if self._task:
            self._task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._task
            self._task = None


def render_sparkline_svg(
    values: list[float],
    width: int = 100,
    height: int = 20,
    stroke_color: str = "#4ade80",
    fill_color: str = "rgba(74, 222, 128, 0.2)",
) -> str:
    """Render a list of values as an SVG sparkline.

    Args:
        values: List of values to plot (0-100 range expected for CPU %)
        width: SVG width in pixels
        height: SVG height in pixels
        stroke_color: Line color
        fill_color: Fill color under the line

    Returns:
        SVG string
    """
    if not values:
        # Empty placeholder
        return f'<svg width="{width}" height="{height}" xmlns="http://www.w3.org/2000/svg"></svg>'

    # Normalize values to 0-1 range
    max_val = max(values) if max(values) > 0 else 1
    normalized = [v / max_val for v in values]

    # Calculate points
    points = []
    x_step = width / max(len(values) - 1, 1)
    for i, v in enumerate(normalized):
        x = i * x_step
        y = height - (v * (height - 2)) - 1  # Leave 1px margin
        points.append(f"{x:.1f},{y:.1f}")

    path_line = " ".join(points)

    # Create filled area path (line + close to bottom)
    fill_points = [*points, f"{width},{height}", f"0,{height}"]
    path_fill = " ".join(fill_points)

    svg = f'''<svg width="{width}" height="{height}" xmlns="http://www.w3.org/2000/svg">
  <polygon points="{path_fill}" fill="{fill_color}" />
  <polyline points="{path_line}" fill="none" stroke="{stroke_color}" stroke-width="1.5" />
</svg>'''
    return svg

File: webterm/docker_stats.go
package webterm

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

const (
	StatsHistorySize = 180
	PollInterval     = 10 * time.Second
)

type DockerStatsCollector struct {
	socketPath     string
	composeProject string

	mu          sync.RWMutex
	cpuHistory  map[string][]float64
	prevCPU     map[string][2]uint64
	serviceList []string

	running bool
	stopCh  chan struct{}
	doneCh  chan struct{}
}

func NewDockerStatsCollector(socketPath, composeProject string) *DockerStatsCollector {
	if socketPath == "" {
		socketPath = DockerSocketPath()
	}
	return &DockerStatsCollector{
		socketPath:     socketPath,
		composeProject: composeProject,
		cpuHistory:     map[string][]float64{},
		prevCPU:        map[string][2]uint64{},
		stopCh:         make(chan struct{}),
		doneCh:         make(chan struct{}),
	}
}

func (d *DockerStatsCollector) Available() bool {
	path := d.socketPath
	if !filepath.IsAbs(path) {
		path = filepath.Clean(path)
	}
	if _, err := os.Stat(path); err != nil {
		return false
	}
	client := newUnixHTTPClient(d.socketPath, 2*time.Second)
	resp, err := client.Get("http://unix/_ping")
	if err != nil {
		return false
	}
	_ = resp.Body.Close()
	return true
}

func (d *DockerStatsCollector) Start(serviceNames []string) {
	d.mu.Lock()
	if d.running {
		d.mu.Unlock()
		return
	}
	d.stopCh = make(chan struct{})
	d.doneCh = make(chan struct{})
	d.serviceList = append([]string{}, serviceNames...)
	d.running = true
	d.mu.Unlock()
	go d.pollLoop()
}

func (d *DockerStatsCollector) Stop() {
	d.mu.Lock()
	if !d.running {
		d.mu.Unlock()
		return
	}
	d.running = false
	close(d.stopCh)
	d.mu.Unlock()
	<-d.doneCh
}

func (d *DockerStatsCollector) AddService(name string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	for _, existing := range d.serviceList {
		if existing == name {
			return
		}
	}
	d.serviceList = append(d.serviceList, name)
}

func (d *DockerStatsCollector) RemoveService(name string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	filtered := d.serviceList[:0]
	for _, item := range d.serviceList {
		if item != name {
			filtered = append(filtered, item)
		}
	}
	d.serviceList = filtered
	delete(d.cpuHistory, name)
	delete(d.prevCPU, name)
}

func (d *DockerStatsCollector) GetCPUHistory(name string) []float64 {
	d.mu.RLock()
	defer d.mu.RUnlock()
	history := d.cpuHistory[name]
	out := make([]float64, len(history))
	copy(out, history)
	return out
}

func (d *DockerStatsCollector) pollLoop() {
	defer close(d.doneCh)
	ticker := time.NewTicker(PollInterval)
	defer ticker.Stop()
	serviceToContainer := map[string]string{}
	refreshCounter := 0

	for {
		select {
		case <-d.stopCh:
			return
		default:
		}

		d.mu.RLock()
		services := append([]string{}, d.serviceList...)
		d.mu.RUnlock()
		if refreshCounter%30 == 0 || len(serviceToContainer) != len(services) {
			serviceToContainer = d.discoverContainers(services)
		}
		refreshCounter++
		for _, service := range services {
			containerID := serviceToContainer[service]
			if containerID == "" {
				continue
			}
			d.pollContainer(service, containerID)
		}
		select {
		case <-d.stopCh:
			return
		case <-ticker.C:
		}
	}
}

func (d *DockerStatsCollector) discoverContainers(serviceNames []string) map[string]string {
	status, body, err := unixJSONRequest(d.socketPath, "GET", "/containers/json", nil)
	if err != nil || status != 200 {
		return map[string]string{}
	}
	var containers []map[string]any
	if err := json.Unmarshal(body, &containers); err != nil {
		return map[string]string{}
	}
	mapping := map[string]string{}
	for _, container := range containers {
		labels := toStringMap(container["Labels"])
		if d.composeProject != "" && labels["com.docker.compose.project"] != d.composeProject {
			continue
		}
		service := labels["com.docker.compose.service"]
		containerID := asString(container["Id"])
		if len(containerID) > 12 {
			containerID = containerID[:12]
		}
		for _, target := range serviceNames {
			if service == target {
				mapping[target] = containerID
				break
			}
			names := toStringSlice(container["Names"])
			for _, name := range names {
				clean := strings.TrimPrefix(name, "/")
				if clean == target || strings.Contains(clean, target) {
					mapping[target] = containerID
					break
				}
			}
		}
	}
	return mapping
}

func (d *DockerStatsCollector) pollContainer(serviceName, containerID string) {
	path := fmt.Sprintf("/containers/%s/stats?stream=false", containerID)
	status, body, err := unixJSONRequest(d.socketPath, "GET", path, nil)
	if err != nil || status != 200 {
		return
	}
	var stats map[string]any
	if err := json.Unmarshal(body, &stats); err != nil {
		return
	}
	cpuStats := toAnyMap(stats["cpu_stats"])
	precpuStats := toAnyMap(stats["precpu_stats"])
	value, ok := d.calculateCPUPercent(serviceName, cpuStats, precpuStats)
	if !ok {
		return
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	history := append(d.cpuHistory[serviceName], value)
	if len(history) > StatsHistorySize {
		history = history[len(history)-StatsHistorySize:]
	}
	d.cpuHistory[serviceName] = history
}

func (d *DockerStatsCollector) calculateCPUPercent(container string, cpuStats, precpuStats map[string]any) (float64, bool) {
	cpuUsage := toAnyMap(cpuStats["cpu_usage"])
	precpuUsage := toAnyMap(precpuStats["cpu_usage"])
	cpuTotal := toUint(cpuUsage["total_usage"])
	preTotal := toUint(precpuUsage["total_usage"])
	systemCPU := toUint(cpuStats["system_cpu_usage"])
	preSystem := toUint(precpuStats["system_cpu_usage"])

	d.mu.Lock()
	if preTotal == 0 {
		if previous, ok := d.prevCPU[container]; ok {
			preTotal = previous[0]
			preSystem = previous[1]
		}
	}
	d.prevCPU[container] = [2]uint64{cpuTotal, systemCPU}
	d.mu.Unlock()

	cpuDelta := int64(cpuTotal) - int64(preTotal)
	systemDelta := int64(systemCPU) - int64(preSystem)
	if cpuDelta < 0 || systemDelta <= 0 {
		return 0, false
	}
	onlineCPUs := toInt(cpuStats["online_cpus"])
	if onlineCPUs <= 0 {
		perCPU := toAnySlice(cpuUsage["percpu_usage"])
		if len(perCPU) == 0 {
			onlineCPUs = 1
		} else {
			onlineCPUs = len(perCPU)
		}
	}
	percent := (float64(cpuDelta) / float64(systemDelta)) * float64(onlineCPUs) * 100.0
	maxValue := float64(onlineCPUs) * 100.0
	return math.Min(percent, maxValue), true
}

func RenderSparklineSVG(values []float64, width, height int) string {
	if width <= 0 {
		width = 100
	}
	if height <= 0 {
		height = 20
	}
	if len(values) == 0 {
		return fmt.Sprintf(`<svg width="%d" height="%d" xmlns="http://www.w3.org/2000/svg"></svg>`, width, height)
	}
	points := make([]float64, 0, len(values)+1)
	points = append(points, 0)
	points = append(points, values...)
	maxValue := 1.0
	for _, value := range points {
		if value > maxValue {
			maxValue = value
		}
	}
	if maxValue <= 0 {
		maxValue = 1
	}
	xStep := float64(width) / float64(max(1, len(points)-1))
	line := strings.Builder{}
	fill := strings.Builder{}
	for i, value := range points {
		x := float64(i) * xStep
		y := float64(height) - ((value / maxValue) * float64(height-2)) - 1
		if i > 0 {
			line.WriteByte(' ')
			fill.WriteByte(' ')
		}
		line.WriteString(fmt.Sprintf("%.1f,%.1f", x, y))
		fill.WriteString(fmt.Sprintf("%.1f,%.1f", x, y))
	}
	fill.WriteString(fmt.Sprintf(" %d,%d 0,%d", width, height, height))
	return fmt.Sprintf(`<svg width="%d" height="%d" xmlns="http://www.w3.org/2000/svg"><polygon points="%s" fill="rgba(74, 222, 128, 0.2)" /><polyline points="%s" fill="none" stroke="#4ade80" stroke-width="1.5" /></svg>`, width, height, fill.String(), line.String())
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func toAnyMap(value any) map[string]any {
	switch raw := value.(type) {
	case map[string]any:
		return raw
	case map[any]any:
		out := map[string]any{}
		for key, val := range raw {
			if text, ok := key.(string); ok {
				out[text] = val
			}
		}
		return out
	default:
		return map[string]any{}
	}
}

func toStringMap(value any) map[string]string {
	out := map[string]string{}
	for key, val := range toAnyMap(value) {
		out[key] = asString(val)
	}
	return out
}

func toAnySlice(value any) []any {
	switch raw := value.(type) {
	case []any:
		return raw
	default:
		return nil
	}
}

func toStringSlice(value any) []string {
	raw := toAnySlice(value)
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		if text, ok := item.(string); ok {
			out = append(out, text)
		}
	}
	return out
}

func toUint(value any) uint64 {
	switch v := value.(type) {
	case uint64:
		return v
	case int:
		if v > 0 {
			return uint64(v)
		}
	case int64:
		if v > 0 {
			return uint64(v)
		}
	case float64:
		if v > 0 {
			return uint64(v)
		}
	case json.Number:
		n, _ := v.Int64()
		if n > 0 {
			return uint64(n)
		}
	}
	return 0
}

func toInt(value any) int {
	switch v := value.(type) {
	case int:
		return v
	case int64:
		return int(v)
	case float64:
		return int(v)
	case json.Number:
		n, _ := v.Int64()
		return int(n)
	}
	return 0
}
